lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,208 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Intel Corporation, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of BitCast to nonstandard types. LLVM
+// transformations bitcast some vector types to scalar types, which are not
+// universally supported across all targets. We need ensure that "optimized"
+// LLVM IR doesn't have primitive types other than supported by the
+// SPIR target (i.e. "scalar 8/16/32/64-bit integer and 16/32/64-bit floating
+// point types, 2/3/4/8/16-element vector of scalar types").
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spv-lower-bitcast-to-nonstandard-type"
+
+#include "SPIRVInternal.h"
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+#include <stack>
+#include <utility>
+
+using namespace llvm;
+
+namespace {
+
+class SPIRVLowerBitCastToNonStandardTypeBase {
+public:
+  SPIRVLowerBitCastToNonStandardTypeBase() {}
+
+  bool runLowerBitCastToNonStandardType(Module &Module) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known
+    auto *M = &Module;
+    bool Changed = false;
+
+    std::vector<std::pair<Instruction *, VectorType *>> BCastsToNonStdVec;
+    std::stack<Instruction *> InstsToErase;
+    for (auto &F : M->functions())
+      for (auto &BB : F)
+        for (auto &I : BB)
+          if (BitCastInst *BC = dyn_cast<BitCastInst>(&I)) {
```

Comment:
You can use auto when type is obvious. + to be consistent with the following code
```suggestion
          if (auto *BC = dyn_cast<BitCastInst>(&I)) {
```